modules/phase_field/include/ics/SmoothCircleFromFileIC.h

```diff
@@ -0,0 +1,47 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef SMOOTHCIRCLEFROMFILE_H
+#define SMOOTHCIRCLEFROMFILE_H
+
+#include "SmoothCircleBaseIC.h"
+#include "DelimitedFileReader.h"
+
+// Forward Declarations
+class SmoothCircleFromFileIC;
+
+template <>
+InputParameters validParams<SmoothCircleFromFileIC>();
+
+/**
+ * Reads multiple circles from a text file with the columns labeled
+ * x   y   z   r. It expects the file to have a one-line header.
+ * Applies all of the circles to the same variable.
+**/
+
+class SmoothCircleFromFileIC : public SmoothCircleBaseIC
+{
+public:
+  SmoothCircleFromFileIC(const InputParameters & parameters);
+
+protected:
+  virtual void computeCircleRadii();
+  virtual void computeCircleCenters();
+
+  std::vector<Real> _x_positions;
+  std::vector<Real> _y_positions;
+  std::vector<Real> _z_positions;
+  std::vector<Real> _input_radii;
+
+  std::vector<std::vector<Real>> _data;
+  std::vector<signed int> _col_map;
```

Comment:
Optional: You could actually make use of the `std::array` type here. Those are for fixed sized "vectors" so you put the size in the declaration with the type and then you don't resize them. You certainly don't have to do this, but this is a good place to use one.